File: Backend/cinema/src/main/java/com/lepham/cinema/exception/ErrorCode.java
package com.lepham.cinema.exception;

import lombok.Getter;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;

import java.net.http.HttpRequest;

@Getter
public enum ErrorCode {
    INVALID_JSON_FORMAT(8888, "Invalid Json format", HttpStatus.BAD_REQUEST),
    UNCATEGORIZED_EXCEPTION(9999, "Uncategorized error", HttpStatus.INTERNAL_SERVER_ERROR),
    INVALID_KEY(1001,"Invalid key", HttpStatus.BAD_REQUEST),
    UNAUTHORIZED(1002, "You do not have permission", HttpStatus.FORBIDDEN),
    INVALID_EMAIL(1003,"Email invalid",HttpStatus.BAD_REQUEST),
    EXISTS_EMAIL(1004,"Email already exists", HttpStatus.BAD_REQUEST),
    INVALID_PHONE(1005,"Phone Number invalid",HttpStatus.BAD_REQUEST),
    INVALID_DOB(1006,"Day of birth invalid",HttpStatus.BAD_REQUEST),
    ACCOUNT_INACTIVE(1007,"Account inactive ", HttpStatus.BAD_REQUEST),
    EXPIRED_TIME_OTP(1008,"Expired time OTP", HttpStatus.BAD_REQUEST),
    EMAIL_PASSWORD_INCORRECT(1009,"Email or password incorrect", HttpStatus.BAD_REQUEST),
    INCORRECT_OTP(1010,"OTP incorrect",HttpStatus.BAD_REQUEST),
    PASSWORD_INVALID(1011,"Password must have at least 8 characters, one capital letter and one special character",HttpStatus.BAD_REQUEST),
    CATEGORY_NOT_FOUND(1012,"Category not found",HttpStatus.BAD_REQUEST),
    FILM_NOT_FOUND(1013,"Film not found",HttpStatus.BAD_REQUEST),
    ROOM_EXISTS(1014,"Room was existed in movie theater",HttpStatus.BAD_REQUEST),
    CATEGORY_NAME_INVALID(1015,"Request has category name is empty ",HttpStatus.BAD_REQUEST),
    STRING_IS_EMPTY(1016,"Information has not been completely filled in",HttpStatus.BAD_REQUEST),
    DURATION_INVALID(1017,"Duration is in 90 to 150 minute",HttpStatus.BAD_REQUEST),
    UNAUTHENTICATED(1018, "Unauthenticated", HttpStatus.UNAUTHORIZED),
    INVALID_DATE(1019, "Invalid date", HttpStatus.BAD_REQUEST),
    NUMBER_NOT_NEGATIVE(1020, "Number is not negative", HttpStatus.BAD_REQUEST),
    INVALID_PRICE(1021, "Price is invalid", HttpStatus.BAD_REQUEST),
    NOT_EXISTS_EMAIL(1022, "Email does not exists", HttpStatus.BAD_REQUEST),
    CATEGORY_NAME_DUPLICATE(1023, "Category name is duplicate", HttpStatus.BAD_REQUEST),
    FILM_NAME_DUPLICATE(1024, "Film title is duplicate", HttpStatus.BAD_REQUEST),
    SEAT_WAS_ORDERED(1025, "Seat was ordered", HttpStatus.BAD_REQUEST),
    VOUCHER_NOT_ENOUGH(1026, "Voucher was not enough", HttpStatus.BAD_REQUEST),
    SEAT_NOT_ORDERED(1027, "Seat was ordered", HttpStatus.OK),
    ACCOUNT_NOT_EXIST(1028,"Account was not exist ", HttpStatus.BAD_REQUEST),
    SHOWTIME_IS_COMING_SOON(1029,"Showtime is coming soon, please choose a showtime other ", HttpStatus.BAD_REQUEST),
    ORDER_NOT_FOUND(1030,"Order not found", HttpStatus.BAD_REQUEST),
    FOOD_NOT_FOUND(1031,"Food not found",HttpStatus.BAD_REQUEST),
    ORDER_NOT_BELONG_ACCOUNT(1032,"The order does not belong to the account", HttpStatus.BAD_REQUEST),
    CAN_NOT_RATING(1032,"You can not rating and feedback", HttpStatus.BAD_REQUEST),
    WAS_RATING(1033,"You was rating and feedback", HttpStatus.BAD_REQUEST),
    COMPLETE_INFORMATION(1034,"Please enter complete information", HttpStatus.BAD_REQUEST),
    DUPLICATE_PASSWORD(1035,"New password must difference old password", HttpStatus.BAD_REQUEST),
    START_TIME_NOT_TODAY(1036,"Start time is not today", HttpStatus.BAD_REQUEST),
    ROOM_NOT_FOUND(1037,"Room not found",HttpStatus.BAD_REQUEST),
    SCHEDULE_NOT_FOUND(1038,"Movie schedule not found",HttpStatus.BAD_REQUEST),
    FILM_NOT_RELEASE(1039,"The movie has not been released yet",HttpStatus.BAD_REQUEST),
    DATE_AFTER_NOW(1040,"Date's schedule must after now 7 days",HttpStatus.BAD_REQUEST),
    ORDER_CAN_NOT_USED(1041,"Order was used or expired",HttpStatus.BAD_REQUEST),
    HOLD_SEAT_ABOVE_LIMIT(1042,"You can only book a maximum of 10 seats",HttpStatus.BAD_REQUEST),
    VOUCHER_NOT_FOUND(1043,"Voucher not found",HttpStatus.BAD_REQUEST),
    STRING_SEAT_INCORRECT(1044,"String seat format is incorrect",HttpStatus.BAD_REQUEST),
    THEATER_NOT_FOUND(1045,"Theater not found",HttpStatus.BAD_REQUEST),
    THEATER_NAME_EXIST(1046,"Theater name is exist",HttpStatus.BAD_REQUEST),
    NOT_TIME_TO_USE(1047,"It's not time to use it yet",HttpStatus.BAD_REQUEST),
    INVALID_FOOD_QUANTITY(1048,"Invalid food quantity",HttpStatus.BAD_REQUEST),
    INCORRECT_PAYMENT_HASH(1049,"Invalid payment hash",HttpStatus.BAD_REQUEST)
    ;


    ErrorCode(int code, String message, HttpStatusCode statusCode) {
        this.code = code;
        this.message = message;
        this.statusCode = statusCode;
    }

    private int code;
    private  String message;
    private HttpStatusCode statusCode;

}

File: App/movie_booking_app/lib/response/error_code.dart
import 'package:flutter_gen/gen_l10n/app_localizations.dart';

class ResponseCode {
  final int code;
  final String message;

  const ResponseCode._(this.code, this.message);

  static ResponseCode nullException(context) {
    return ResponseCode._(8888, AppLocalizations.of(context)!.code_8888);
  }

  static ResponseCode uncategorizedException(context) {
    return ResponseCode._(9999, AppLocalizations.of(context)!.code_9999);
  }

  static ResponseCode invalidKey(context) {
    return ResponseCode._(1001, AppLocalizations.of(context)!.code_1001);
  }

  static ResponseCode unauthorized(context) {
    return ResponseCode._(1002, AppLocalizations.of(context)!.code_1002);
  }

  static ResponseCode invalidEmail(context) {
    return ResponseCode._(1003, AppLocalizations.of(context)!.code_1003);
  }

  static ResponseCode existsEmail(context) {
    return ResponseCode._(1004, AppLocalizations.of(context)!.code_1004);
  }

  static ResponseCode invalidPhone(context) {
    return ResponseCode._(1005, AppLocalizations.of(context)!.code_1005);
  }

  static ResponseCode invalidDob(context) {
    return ResponseCode._(1006, AppLocalizations.of(context)!.code_1006);
  }

  static ResponseCode accountInactive(context) {
    return ResponseCode._(1007, AppLocalizations.of(context)!.code_1007);
  }

  static ResponseCode expiredTimeOtp(context) {
    return ResponseCode._(1008, AppLocalizations.of(context)!.code_1008);
  }

  static ResponseCode emailPasswordIncorrect(context) {
    return ResponseCode._(1009, AppLocalizations.of(context)!.code_1009);
  }

  static ResponseCode incorrectOtp(context) {
    return ResponseCode._(1010, AppLocalizations.of(context)!.code_1010);
  }

  static ResponseCode passwordInvalid(context) {
    return ResponseCode._(1011, AppLocalizations.of(context)!.code_1011);
  }

  static ResponseCode categoryNotFound(context) {
    return ResponseCode._(1012, AppLocalizations.of(context)!.code_1012);
  }

  static ResponseCode filmNotFound(context) {
    return ResponseCode._(1013, AppLocalizations.of(context)!.code_1013);
  }

  static ResponseCode roomExists(context) {
    return ResponseCode._(1014, AppLocalizations.of(context)!.code_1014);
  }

  static ResponseCode categoryNameInvalid(context) {
    return ResponseCode._(1015, AppLocalizations.of(context)!.code_1015);
  }

  static ResponseCode stringIsEmpty(context) {
    return ResponseCode._(1016, AppLocalizations.of(context)!.code_1016);
  }

  static ResponseCode durationInvalid(context) {
    return ResponseCode._(1017, AppLocalizations.of(context)!.code_1017);
  }

  static ResponseCode unauthenticated(context) {
    return ResponseCode._(1018, AppLocalizations.of(context)!.code_1018);
  }

  static ResponseCode invalidDate(context) {
    return ResponseCode._(1019, AppLocalizations.of(context)!.code_1019);
  }

  static ResponseCode numberNotNegative(context) {
    return ResponseCode._(1020, AppLocalizations.of(context)!.code_1020);
  }

  static ResponseCode invalidPrice(context) {
    return ResponseCode._(1021, AppLocalizations.of(context)!.code_1021);
  }

  static ResponseCode notExistsEmail(context) {
    return ResponseCode._(1022, AppLocalizations.of(context)!.code_1022);
  }

  static ResponseCode categoryNameDuplicate(context) {
    return ResponseCode._(1023, AppLocalizations.of(context)!.code_1023);
  }

  static ResponseCode filmNameDuplicate(context) {
    return ResponseCode._(1024, AppLocalizations.of(context)!.code_1024);
  }

  static ResponseCode seatWasOrdered(context) {
    return ResponseCode._(1025, AppLocalizations.of(context)!.code_1025);
  }

  static ResponseCode voucherNotEnough(context) {
    return ResponseCode._(1026, AppLocalizations.of(context)!.code_1026);
  }

  static ResponseCode seatNotOrdered(context) {
    return ResponseCode._(1027, AppLocalizations.of(context)!.code_1027);
  }

  static ResponseCode accountNotExist(context) {
    return ResponseCode._(1028, AppLocalizations.of(context)!.code_1028);
  }

  static ResponseCode showtimeIsComingSoon(context) {
    return ResponseCode._(1029, AppLocalizations.of(context)!.code_1029);
  }

  static ResponseCode orderNotFound(context) {
    return ResponseCode._(1030, AppLocalizations.of(context)!.code_1030);
  }

  static ResponseCode foodNotFound(context) {
    return ResponseCode._(1031, AppLocalizations.of(context)!.code_1031);
  }

  static ResponseCode orderNotBelongAccount(context) {
    return ResponseCode._(1032, AppLocalizations.of(context)!.code_1032);
  }

  static ResponseCode cannotRating(context) {
    return ResponseCode._(1032, AppLocalizations.of(context)!.code_1032);
  }

  static ResponseCode wasRating(context) {
    return ResponseCode._(1033, AppLocalizations.of(context)!.code_1033);
  }

  static ResponseCode completeInformation(context) {
    return ResponseCode._(1034, AppLocalizations.of(context)!.code_1034);
  }

  static ResponseCode duplicatePassword(context) {
    return ResponseCode._(1035, AppLocalizations.of(context)!.code_1035);
  }

  static ResponseCode startTimeNotToday(context) {
    return ResponseCode._(1036, AppLocalizations.of(context)!.code_1036);
  }

  static ResponseCode roomNotFound(context) {
    return ResponseCode._(1037, AppLocalizations.of(context)!.code_1037);
  }

  static ResponseCode scheduleNotFound(context) {
    return ResponseCode._(1038, AppLocalizations.of(context)!.code_1038);
  }

  static ResponseCode filmNotRelease(context) {
    return ResponseCode._(1039, AppLocalizations.of(context)!.code_1039);
  }

  static ResponseCode dateAfterNow(context) {
    return ResponseCode._(1040, AppLocalizations.of(context)!.code_1040);
  }

  static ResponseCode orderCannotUsed(context) {
    return ResponseCode._(1041, AppLocalizations.of(context)!.code_1041);
  }

  static ResponseCode holdSeatAboveLimit(context) {
    return ResponseCode._(1042, AppLocalizations.of(context)!.code_1042);
  }

  static ResponseCode voucherNotFound(context) {
    return ResponseCode._(1043, AppLocalizations.of(context)!.code_1043);
  }

  static ResponseCode stringSeatIncorrect(context) {
    return ResponseCode._(1044, AppLocalizations.of(context)!.code_1044);
  }

  static final Map<int, Function> _responseCodeMap = {
    8888: nullException,
    9999: uncategorizedException,
    1001: invalidKey,
    1002: unauthorized,
    1003: invalidEmail,
    1004: existsEmail,
    1005: invalidPhone,
    1006: invalidDob,
    1007: accountInactive,
    1008: expiredTimeOtp,
    1009: emailPasswordIncorrect,
    1010: incorrectOtp,
    1011: passwordInvalid,
    1012: categoryNotFound,
    1013: filmNotFound,
    1014: roomExists,
    1015: categoryNameInvalid,
    1016: stringIsEmpty,
    1017: durationInvalid,
    1018: unauthenticated,
    1019: invalidDate,
    1020: numberNotNegative,
    1021: invalidPrice,
    1022: notExistsEmail,
    1023: categoryNameDuplicate,
    1024: filmNameDuplicate,
    1025: seatWasOrdered,
    1026: voucherNotEnough,
    1027: seatNotOrdered,
    1028: accountNotExist,
    1029: showtimeIsComingSoon,
    1030: orderNotFound,
    1031: foodNotFound,
    1032: orderNotBelongAccount,
    1033: wasRating,
    1034: completeInformation,
    1035: duplicatePassword,
    1036: startTimeNotToday,
    1037: roomNotFound,
    1038: scheduleNotFound,
    1039: filmNotRelease,
    1040: dateAfterNow,
    1041: orderCannotUsed,
    1042: holdSeatAboveLimit,
    1043: voucherNotFound,
    1044: stringSeatIncorrect,
  };

  static ResponseCode? getMessage(int code, context) {
    final responseFunction = _responseCodeMap[code];
    if (responseFunction != null) {
      return responseFunction(context);
    } else {
      return uncategorizedException(context);
    }
  }
}
